### DISCUSSION TITLE
How does trpc-query generate query keys?

### DISCUSSION THREAD

**Discussion by HugeLetters** [trpc/trpc#4920] [Q&A] [answered]
I'm making something similar in a project and I was curious how does trpc generate query keys?

I'm guessing it takes specified router path and then input and just uses that as keys? Basically `[...path[], input]`, right? 
```ts
const query = trpc.user.comments.getAll.useQuery({id: string, before: Date})
// will have that as a key
const queryKey = ['user', 'comments', 'getAll', {id: string, before: Date}] 
```

**Comment by zirkelc** [comment]
Same question here. It would be good if we could somehow influence the generated query key. For example, I have a parameter in the input that acts as as `force refresh` that I would like to send to the server but not have it my query key.

**Reply by HugeLetters** [reply]
Pretty sure you can influence generated key using `queryKeyHashFn` option but be mindful it's pretty low-level in terms of how library works - make sure you generate the key in the same vein as trpc-query does...which brings us back to mine question 😅

This option is a method which receives your query data as an input which by default tanstack-query hashes using their algorithm. You just need to remove some part of the input you don't want to be there and then hash using the same algorithm.

**Reply by zirkelc** [reply]
Oh great, I didn't know about this function! Thanks for the quick reply!

What is strange though is that if I override `queryKeyHashFn` on just one `useQuery` call, it seems to apply to all queries. It behaves as if I set this function on the `QueryClient`.

**Reply by zirkelc** [reply, marked as answer]
Regarding your question, don't know if it helps you but here's is the implementation of tRPC `geQueryKey`:

[getQueryKey](https://github.com/trpc/trpc/blob/b745c5060891675078d4b85fcea919e221d36f33/packages/react-query/src/internals/getQueryKey.ts#L82-L100)
```ts
export function getQueryKey<
  TProcedureOrRouter extends
    | AnyMutationProcedure
    | AnyQueryProcedure
    | AnyRouter,
  TPath extends string,
  TFlags,
>(..._params: GetQueryKeyParams<TProcedureOrRouter, TPath, TFlags>) {
  const [procedureOrRouter, input, type] = _params;
  // @ts-expect-error - we don't expose _def on the type layer
  const path = procedureOrRouter._def().path as string[];
  const dotPath = path.join('.');
  const queryKey = getArrayQueryKey(
    getQueryKeyInternal(dotPath, input),
    type ?? 'any',
  );
  return queryKey;
}

export function getQueryKeyInternal(
  path: string,
  input: unknown,
): [string, unknown] | [string] {
  if (path.length) return input === undefined ? [path] : [path, input];
  return [] as unknown as [string];
}

export function getArrayQueryKey(
  queryKey: unknown[] | string | [string, ...unknown[]] | [string],
  type: QueryType,
): QueryKey {
  const queryKeyArrayed = Array.isArray(queryKey) ? queryKey : [queryKey];
  const [path, input] = queryKeyArrayed;

  const arrayPath =
    typeof path !== 'string' || path === '' ? [] : path.split('.');

  // Construct a query key that is easy to destructure and flexible for
  // partial selecting etc.
  // https://github.com/trpc/trpc/issues/3128
  if (!input && (!type || type === 'any'))
    // for `utils.invalidate()` to match all queries (including vanilla react-query)
    // we don't want nested array if path is empty, i.e. `[]` instead of `[[]]`
    return arrayPath.length ? [arrayPath] : ([] as unknown as QueryKey);
  return [
    arrayPath,
    {
      ...(typeof input !== 'undefined' && { input: input }),
      ...(type && type !== 'any' && { type: type }),
    },
  ];
}
```

**Reply by HugeLetters** [reply]
> What is strange though is that if I override queryKeyHashFn on just one useQuery call, it seems to apply to all queries. It behaves as if I set this function on the QueryClient.


I did a local test - doesn't seem to be the case in my scenario. Here's a code snippet
```tsx
import { api } from "@/utils/api";

export default function Page() {
  const query1 = api.review.getReviewCount.useQuery(undefined, {});
  const query2 = api.review.getReviewCount.useQuery(undefined, {
    queryKeyHashFn() {
      return "hash-is-unique";
    },
  });
  const query3 = api.review.getReviewCount.useQuery(undefined, {});

  return <div>{query1.data} - {query2.data} - {query3.data}</div>
}
```
Which then produces two separate requests - first and third queries have the same hash so same request, 2nd query is a separate request.
![image](https://github.com/trpc/trpc/assets/119697239/060d1e87-7214-4ec4-8cc6-a39d6feaec70)
Perhaps you're confused by the fact that it's single request in a network tab - that's because trpc batches several sync requests together. So two requests are batched together by trpc into a single network call and then processed as 2 queries by tanstack. If all 3 queries had the same queryKey then in network response preview tab I would see only 1 result, not two.

Here's the same scenario but now query2 has hash function removed
![image](https://github.com/trpc/trpc/assets/119697239/cfc5b095-4020-4c70-9f5c-b07d93779339)

**Reply by HugeLetters** [reply]
> Regarding your question, don't know if it helps you but here's is the implementation of tRPC geQueryKey:

Yup, seems like my initial conjecture was true - I've observed the same just now by looking at the input type in the `queryKeyHashFn` function.

**Comment by zirkelc** [comment]
> Here's the same scenario but now query2 has hash function removed ![image](https://user-images.githubusercontent.com/119697239/277679660-cfc5b095-4020-4c70-9f5c-b07d93779339.png)

@HugeLetters  Can you add a `console.log("queryKeyHashFn", input);` to your `queryHashFn()`:

```ts
import { api } from "@/utils/api";

export default function Page() {
  const query1 = api.review.getReviewCount.useQuery(undefined, {});
  const query2 = api.review.getReviewCount.useQuery(undefined, {
    queryKeyHashFn(input) {
      console.log("queryKeyHashFn", input);
      return "hash-is-unique";
    },
  });
  const query3 = api.review.getReviewCount.useQuery(undefined, {});

  return <div>{query1.data} - {query2.data} - {query3.data}</div>
}
```

It prints lots of different `inputs`, especially inputs from other `useQuery` calls.

**Reply by zirkelc** [reply]
Here is an example: https://codesandbox.io/p/sandbox/github/trpc/trpc/tree/main/examples/next-minimal-starter

It prints `queryKeyHasFn` for all queries which are called after the second query:
![image](https://github.com/trpc/trpc/assets/950244/1c450e4a-3299-490c-8466-402e8345d1a7)

**Reply by HugeLetters** [reply]
> Can you add a console.log("queryKeyHashFn", input); to your queryHashFn():

I've replaced undefined with 1,2,3 to differentiate queries - it just prints `[ [ "review", "getReviewCount" ],  { "input": 2, "type": "query" } ]` two times(guess because it's strict mode) for every render.

I'll go check-out your sandbox now

**Reply by HugeLetters** [reply]
> It prints `queryKeyHasFn` for all queries which are called after the second query

![image](https://github.com/trpc/trpc/assets/119697239/1f26c018-2b2f-4796-bc09-3245955fb3bf)
```tsx
  const result1 = trpc.greeting.useQuery({ name: 'one' });
  const result2 = trpc.greeting.useQuery({ name: 'two' }, {
    queryKeyHashFn(input) {
      console.log("queryKeyHashFn", input);
      return "hash-is-unique";
    },
  });
  const result3 = trpc.greeting.useQuery({ });
```

Here's a link to my fork of a [sandbox](https://codesandbox.io/p/sandbox/great-morning-zz8yyn?file=%2Fsrc%2Fpages%2Findex.tsx%3A21%2C33) - perhaps you can point out what am I doing differently. I didn't change any code except for index.tsx

**Reply by HugeLetters** [reply]
Oh, I see - it prints the logs in codesandbox console but not in a browser console. Perhaps just a bug with codesandbox?

Because for sure actually each query has its own separate key hash - otherwise it wouldn't render 'hello world' since it would have the same key hash with query2 and would just render its results.

**Reply by zirkelc** [reply]
Okay, I'll just leave it for now. I've found a workaround with `queryClient.setQueriesData()` and using a partial query key.
